### PR TITLE
ct: move WITH options to a consistent location in the syntax

### PR DESCRIPTION
**Pull request by danhhz**
This allows us to pull more common code out in `AstDisplay`, which will hopefully help prevent bugs like MaterializeInc/database-issues#8719.

Touches https://github.com/MaterializeInc/database-issues/issues/8427

### Motivation

   * This PR refactors existing code.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/sql-parser/src/parser.rs [hunk 1]
```diff
@@ -3605,20 +3605,37 @@ impl<'a> Parser<'a> {
         };
         let in_cluster = self.parse_optional_in_cluster()?;
```
Comment (danhhz) [inline src/sql-parser/src/parser.rs:3606]: Also, there's some opportunity for dedup in the parser, too https://github.com/MaterializeInc/materialize/pull/30260/files/93cb32bdaf21890e65f822711d62f0aec1556af4#diff-ff4ad701e3b476ee7c4794f262a310147cb7448f5da48c590ea9b5648aa89ff2

Comment (danhhz) [reply src/sql-parser/src/parser.rs:3612]: I tried a more aggressive version of this where I pulled out a full prelude/postlude that was shared between all three forms, but IMO it felt a little clunky and didn't seem to get us a ton. Happy to defer to adapter folks though

**Review by danhhz — COMMENTED** [review]
Okay this is ready to go!

File: src/sql-parser/src/parser.rs [hunk 2]
```diff
@@ -3610,20 +3610,22 @@ impl<'a> Parser<'a> {
...
-        } else {
-            vec![]
+        // Also try to parse WITH options in the old location. We never exposed
+        // this to users, so this can be removed once the CI upgrade tests have
+        // moved past these old versions.
+        let legacy_with_options = self.parse_create_continual_task_with_options()?;
+        let with_options = match (!with_options.is_empty(), !legacy_with_options.is_empty()) {
+            (_, false) => with_options,
+            (false, true) => legacy_with_options,
+            (true, true) => todo!("WIP return parser error"),
```
Comment (danhhz) [inline src/sql-parser/src/parser.rs:3628]: Oh except for this, whoops. Fix inc

**Review by danhhz — COMMENTED** [review]
Oh and Nightlies seems happy with my backward compatibility: https://buildkite.com/materialize/nightly/builds/10232

**Review by jkosh44 — APPROVED** [review]
LGTM

**Comment by danhhz** [comment]
TFTR!

**Pull request merged**